dnf/plugins/download/dnf-command-download.c

```diff
@@ -0,0 +1,324 @@
+    }
+  }
+  return TRUE;
+}
+
+static gint
+gptrarr_dnf_package_repopkgcmp (gconstpointer a, gconstpointer b)
+{
+  // Sort by repo first, then by package
+  // so we only need to add repo keys the first time we see a repo
+  // in the package list
+  DnfPackage **apkg = (DnfPackage**)a;
+  DnfPackage **bpkg = (DnfPackage**)b;
+  const gchar * areponame = dnf_package_get_reponame (*apkg);
+  const gchar * breponame = dnf_package_get_reponame (*bpkg);
+  int repocmp = strcmp (areponame, breponame);
+  if (repocmp == 0) {
+    return dnf_package_cmp (*apkg, *bpkg);
+  }
+  return repocmp;
+}
+
+static gboolean
+dnf_command_download_run (DnfCommand      *cmd,
+                          int              argc,
+                          char            *argv[],
+                          GOptionContext  *opt_ctx,
+                          DnfContext      *ctx,
+                          GError         **error)
+{
+  g_auto(GStrv) opt_key = NULL;
+  gboolean opt_src = FALSE;
+  const GOptionEntry opts[] = {
+    { "source", '\0', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &opt_src, "download source packages", NULL },
+    { G_OPTION_REMAINING, '\0', 0, G_OPTION_ARG_STRING_ARRAY, &opt_key, NULL, NULL },
+    { NULL }
+  };
+  g_option_context_add_main_entries (opt_ctx, opts, NULL);
+
+  if (!g_option_context_parse (opt_ctx, &argc, &argv, error))
+    return FALSE;
+
+  DnfRepoLoader * repo_loader = dnf_context_get_repo_loader (ctx);
+  // If -source arg was passed, auto-enable the source repositories
+  if (opt_src) {
+    if (!dnf_command_download_enablesourcerepos (ctx, repo_loader, error)) {
+      return FALSE;
+    }
+  }
+
+  if (opt_key == NULL)
+    {
+      g_set_error_literal (error,
+                           G_IO_ERROR,
+                           G_IO_ERROR_FAILED,
+                           "Packages are not specified");
+      return FALSE;
+    }
+
+  DnfState * state = dnf_context_get_state (ctx);
+  DnfContextSetupSackFlags sack_flags = DNF_CONTEXT_SETUP_SACK_FLAG_SKIP_RPMDB;
+  dnf_context_setup_sack_with_flags (ctx, state, sack_flags, error);
+  DnfSack *sack = dnf_context_get_sack (ctx);
+
+  hy_autoquery HyQuery query = hy_query_create (sack);
+
+  if (opt_key)
+    {
+      hy_query_filter_empty (query);
+      for (char **pkey = opt_key; *pkey; ++pkey)
+        {
+          g_auto(HySubject) subject = hy_subject_create (*pkey);
+          HyNevra out_nevra;
+          hy_autoquery HyQuery key_query =
+            hy_subject_get_best_solution (subject, sack, NULL, &out_nevra,
+                                          TRUE, TRUE, FALSE, TRUE, opt_src);
+          if (out_nevra)
+            {
+              hy_nevra_free (out_nevra);
+            }
+          hy_query_union (query, key_query);
+        }
+    }
+  if (opt_src) {
+    hy_query_filter (query, HY_PKG_ARCH, HY_EQ, "src");
+  }
+  else {
+    hy_query_filter (query, HY_PKG_ARCH, HY_NEQ, "src");
+  }
+  hy_query_filter_latest_per_arch (query, TRUE);
```

Comment:
1. DNF `download` command takes repository priority into account.
Just replace the filter with:
`hy_query_filter_num (query, HY_PKG_LATEST_PER_ARCH_BY_PRIORITY, HY_EQ, 1);`

2. Move the filter line up (to the `for` loop) before the `hy_query_union (query, key_query);` line and replace `query` by `key_query`.
It is necessary to support the download of multiple versions of the package explicitly requested by the user - E.g. `microdnf download microdnf-3.6.0 microdnf-3.4.0`